jvm-agent/src/main/org/jetbrains/lincheck/jvm/agent/analysis/controlflow/BasicBlockControlFlowGraph.kt

```diff
@@ -97,20 +114,100 @@ class BasicBlockControlFlowGraph(
         val node = instructions.get(idx)
         return (node as? LabelNode)?.label
     }
-    
+
     /**
      * Computes loop-related information for this method.
      */
     fun computeLoopInformation(): MethodLoopsInformation {
         if (loopInfo == null) {
-            // TODO: this is a stub: loop detection is not implemented yet; the method returns empty sites.
-            loopInfo = MethodLoopsInformation().apply {
-                loops.forEach { it.validateLoopEdgesInvariants() }
+            // compute predecessors for all basic blocks
+            val allPredecessors: Array<MutableSet<BasicBlockIndex>> = Array(basicBlocks.size) { mutableSetOf() }
+            val normalPredecessors: Array<MutableSet<BasicBlockIndex>> = Array(basicBlocks.size) { mutableSetOf() }
+            for (e in edges) {
+                val u = e.source
+                val v = e.target
+                allPredecessors[v].add(u)
+                if (e.label !is EdgeLabel.Exception) {
+                    normalPredecessors[v].add(u)
+                }
+            }
+            // compute loop information
+            val dominators = computeDominators(allPredecessors)
+            if (isReducible(dominators)) {
+                _isReducible = true
+                // CFG is reducible, so we can compute actual loops
+                loopInfo = computeLoopsFromDominators(dominators, normalPredecessors).also { info ->
+                    info.validateBasicBlocksLoopsMapping()
+                    info.loops.forEach {
+                        it.validateLoopEdgesInvariants()
+                    }
+                }
+            }
+            else {
+                _isReducible = false
+                // CFG is irreducible, our loop calculation algorithm will not work with it
+                // so we report that fact and don't compute any loop information
+                Logger.warn { "Irreducible CFG detected, loop information will not be computed:\n${toFormattedString()}" }
+                loopInfo = MethodLoopsInformation()
             }
         }
         return loopInfo!!
     }
 
+    /**
+     * CFG is called *reducible* if removal of its back edges leads to a graph that:
+     * * is acyclic (in terms of graph theory)
+     * * each basic block of CFG can be reached from the initial one
+     *
+     * The initial basic block in our case is the basic block at index 0.
+     * And reachability calculation can use both normal (except for back-edges) and exceptional edges.
+     *
+     * See also https://www.csd.uwo.ca/~mmorenom/CS447/Lectures/CodeOptimization.html/node6.html.
+     */
+    private fun isReducible(dominators: Array<Set<BasicBlockIndex>>): Boolean {
```

Comment:
Let's move to `LoopInformation.kt` --- it is conceptually related to other methods that compute loop information.